loggers/logger.ts

```diff
@@ -0,0 +1,191 @@
+import { colors } from "../deps.ts";
+const encoder = new TextEncoder();
+
+export interface LoggerConfigs {
+  // deno-lint-ignore camelcase
+  tag_string?: string;
+  tag_string_fns?: { [key: string]: any }; // `any` because it can be a string, or an object with functions and/or strings, and the compiler throws errors when trying to execute certain logic with said type
+}
+
+type LogTypes = "info" | "debug" | "warn" | "error";
+
+/**
+ * This Logger is the base logger class for all logger classes.
+ */
+export abstract class Logger {
+  /**
+       * @param configs - Config used for Logging
+       */
+  protected configs: LoggerConfigs;
+
+  /**
+       * The level of the log message currently being written.
+       */
+  protected current_log_message_level_name = "";
+
+  //////////////////////////////////////////////////////////////////////////////
+  // FILE MARKER - CONSTRUCTOR /////////////////////////////////////////////////
+  //////////////////////////////////////////////////////////////////////////////
+
+  /**
+       * Construct an object of this class.
+       *
+       * @param configs - Config used for Logging
+       */
+  constructor(configs: LoggerConfigs) {
+    if (!configs.tag_string) {
+      configs.tag_string = "";
+    }
+
+    if (!configs.tag_string_fns) {
+      configs.tag_string_fns = {};
+    }
+
+    this.configs = configs;
+  }
+
+  //////////////////////////////////////////////////////////////////////////////
+  // FILE MARKER - METHODS - ABSTRACT //////////////////////////////////////////
+  //////////////////////////////////////////////////////////////////////////////
+
+  /**
+       * Write a message to the console. Prefixed with the log type
+       *
+       * @param message - The message to be logged
+       *
+       * @returns Return the full logged message.
+       */
+  abstract debug(message: string): string;
+
+  /**
+       * Write a message to the console. Prefixed with the log type
+       *
+       * @param message - The message to be logged
+       *
+       * @returns Return the full logged message.
+       */
+  abstract info(message: string): string;
+
+  /**
+       * Write a message to the console. Prefixed with the log type
+       *
+       * @param message - The message to be logged
+       *
+       * @returns Return the full logged message.
+       */
+  abstract warn(message: string): string;
+
+  /**
+       * Write a message to the console. Prefixed with the log type
+       *
+       * @param message - The message to be logged
+       *
+       * @returns Return the full logged message.
+       */
+  abstract error(message: string): string;
+
+  //////////////////////////////////////////////////////////////////////////////
+  // FILE MARKER - METHODS - PROTECTED /////////////////////////////////////////
+  //////////////////////////////////////////////////////////////////////////////
+
```

Comment:
i think this base `Logger` class shouldn't know about how to log to the console or write to files. those should be in the console or file logger classes